### DISCUSSION TITLE
I am losing my mind with ctlout and midiout

### DISCUSSION THREAD

**Discussion by jocphx** [pierreguillot/Camomile#260] [Q&A]
Hi,
I've spent a good deal of time trying to figure out how to use Camomile to create a MIDI function generator, to use as a modulation source in a DAW. Nothing that I tried so far seems to have any effect.
The bigger picture would be to create clock independent generators and modulators, often called "free running".
I am on Win10, using FL Studio 20 and Ableton 10.

In each DAW, I send the MIDI output of the patch to a synth (I am using Surge atm.)

I am attaching the basic proof-of-concept patch. The [noteout] works fine. I have no idea how to check that [midiout] and/or [ctlout] even work. 
I was attempting to control, say, one of Surge's oscillators volume with either CC 20 or CC 80.

All I could find online, are years old entries, in various forums, mentioning that those objects work only on Linux.

Any help, suggestion is greatly appreciated!

![Capture](https://user-images.githubusercontent.com/76123145/117551076-b2b6b280-aff8-11eb-85c3-ed8b4b695cf8.JPG)

**Comment by pierreguillot** [comment]
Hi,

I managed to output MIDI CC, to test it I created another plugin with [ctlin] and [print] that I put after the MIDI CC generator. 

Which version of Camomile do you use? And which format (VST2 or VST3)?

And there have been issues with MIDI CC (https://github.com/pierreguillot/Camomile/issues?q=is%3Aissue+cc+), maybe these threads might help you.

**Reply by jocphx** [reply]
Hi @pierreguillot ,

I was using v 1.0.7 as VST3. 

I followed your suggestion, and searched among the existing issues, and this caught my attention: https://github.com/pierreguillot/Camomile/issues/150 

I replaced my 1.0.7 VST3 with the 1.0.6 VST2 and the patch's [ctlout] started to work!

So, it looks like *the issue is still present in 1.0.7 and 1.0.8-beta4* (I just tested that) since both offer only VST3, on Windows 10.

Shall I open a new issue?


As the patch started to work, I tested also the [midiout] section. I modified the message to be [176 20 $1( , if I am not mistaken, it should behave identically to the [ctlout 20 1], but I observed that the Surge's slider that I attached to CC20, instead of moving smoothly, it jumps up and down erratically, as I change the value of the number box.

This begs the question, am I using the [midiout] object incorrectly or is this possibly a bug?

Please let me know if there are any other tests that you would like me to run.

**Reply by pierreguillot** [reply]
Hi,
I'm sorry for this late answer. If I remember well (and it seems to be confirmed by the 1.0.7 changelog), I fixed the [midiout] object between 1.0.6 and 1.0.7. But as I also dropped the VST2 format and the midi cc support changed with the VST3 format, this doesn't help your situation... I think the issue is related to https://github.com/pierreguillot/Camomile/issues/194. You should also read this issue https://github.com/falkTX/Carla/issues/1240.

**Reply by jocphx** [reply]
Yes, it looks like CC send is broken with VST3. 

You are mentioning fixing [midiout], but the tests I ran show otherwise, since the original patch I uploaded at the beginning of this thread was using either [midiout] or [ctlout], hence my question: am I using [midiout] wrong?

Juce 6.1 came out on Aug. 23 2021, that can possibly fix this. 

"VST3 Plug-in Extensions
JUCE's VST3 backend has been redesigned, offering greater compatibility with the VST3 specification and enabling access to VST3-specific extensions. Plug-in developers can now access a much richer interface to any DAWs that provide their own set of extensions, such as those published by Presonus, Cockos and others."

**Comment by pierreguillot** [comment]
Okay, I'll update Juce then. Thank you!

**Reply by pierreguillot** [reply]
I updated Juce, you can find the binaries here https://github.com/pierreguillot/Camomile/actions/runs/1402491699.

**Reply by jocphx** [reply]
Hi @pierreguillot , unfortunately MIDI CC is still broken in VST3.

I've done some more reading, and it looks like VST3 had broken MIDI, everywhere.

https://forums.steinberg.net/t/vst3-and-midi-cc-pitfall/201879/8
https://www.kvraudio.com/forum/viewtopic.php?t=538083
https://www.kvraudio.com/forum/viewtopic.php?t=540368
https://forum.juce.com/t/midi-cc-messages-in-vst3/34031/3

Was it a VST licensing issue that led you to abandon VST2? I am asking because MIDI CC works fine in VST2, and not at all in VST3.

Last but not least, thank you for your hard work!

**Reply by pierreguillot** [reply]
What I understand about MIDI CC is that it is complicated to manage on the host side because when using MIDI CC to change something in the plugin, the host is not aware of the change - and I think that's why VST3 dropped (partially) the MIDI CC support. And it seems that some DAWs offer the possibility to link the MIDI CC to some plugin parameters (so the link mechanism is managed by the DAW and not by the plugin itself) and the VST3 now also introduces a mechanism to link internal the MIDI CC to plugin parameters. And, I think, all these things are working well. So I don't think that MIDI is broken in VST3 but just different to ensure better stability and consistency between the different implementations. 

And if we suppose that what I wrote is not wrong that should be logical to prevent plugins to send MIDI CC. 

And if we suppose that what I wrote is not wrong that should be logical to prevent plugins to send MIDI CC. And I think if this is the case, there might be a really good reason (I doubt that they just want to bother people). I know it doesn't help sorry... 

What do you want to achieve by the way with MIDI CC? Perhaps using OSC would do the job or simply using Pure Data vanilla as a MIDI controller?

About VST2, here is the issue: https://github.com/pierreguillot/Camomile/issues/205.

**Reply by jocphx** [reply]
I would like to use MIDI CC to send my own kind of random control signals to the different VST I am using in a project. 
I already created a few generative sequencers in PD/Camomile, therefore Note and Velocity MIDI messages work fine. I wanted to add random faders and panners.

The reason I started to work on this project is that there are not many existing VST plugins that generate MIDI CC, and they do not have generative capacity I am looking for, so I thought to make my own. 

Working directly from PD would be a hindrance, since I'll have to save 2 separate projects, one in the DAW the other in PD, and remember which one goes with which one. I tried to use other software that way, it just did not flow well.

I honestly do not understand why, whomever wrote the VST3 spec, did it this way.... BTW my DAW is FLStudio, in case you might have some secret sauce for it.
